### PR TITLE
[RFC] Use floating window as autocmd window

### PR DESCRIPTION
**Pull request by mhinz**
These are just minimal changes to make it work with a floating window so far. A lot of assumptions can now be purged from the remaining code.

### CODE REVIEW

File: src/nvim/fileio.c [hunk 1]
```diff
@@ -6558,16 +6558,17 @@ aucmd_prepbuf (
...
     save_ea = p_ea;
     p_ea = false;
 
     /* Prevent chdir() call in win_enter_ext(), through do_autochdir(). */
     save_acd = p_acd;
     p_acd = false;
 
-    (void)win_split_ins(0, WSP_TOP, aucmd_win, 0);
-    (void)win_comp_pos();       /* recompute window positions */
+    win_enter(aucmd_win, false);
```
Comment (bfredl) [inline src/nvim/fileio.c]: if the aucmd window is not new `win_append` is needed here, to match the `win_remove`. Also I don't think `win_enter` should be used (it is only to change the real current window).

File: src/nvim/fileio.c [hunk 2]
```diff
@@ -6615,7 +6616,7 @@ void aucmd_restbuf(aco_save_T *aco)
       close_tabpage(curtab);
     }
 
-    restore_snapshot(SNAP_AUCMD_IDX, false);
+    /* restore_snapshot(SNAP_AUCMD_IDX, false); */
     (void)win_comp_pos();       // recompute window positions
```
Comment (bfredl) [inline src/nvim/fileio.c]: this should also be removed I think (as window layout did not change)

Comment (mhinz) [reply src/nvim/fileio.c]: Yup, already removed in my local copy!

**Comment by justinmk** [comment]
May be useful: https://github.com/vim/vim/commit/a68e59590905da9b4448ff1fcac929ad1a18da9e

**Comment by justinmk** [comment]
This helps significantly with #8723 (which is still a problem with e.g. 'diff' mode).

@mhinz  What's blocking this?

**Comment by mhinz** [comment]
ASAN complains about leaking bytes but I didn't find the culprit. I'll have another look.

**Comment by mhinz** [comment]
@justinmk Should be reviewable now. Maybe you can already check if it helps with #8723.

File: src/nvim/fileio.c [hunk 3]
```diff
@@ -6552,6 +6545,7 @@ aucmd_prepbuf (
   /* Allocate "aucmd_win" when needed. */
   if (win == NULL && aucmd_win == NULL) {
     win_alloc_aucmd_win();
+    auwin_appended = true;
```
Comment (justinmk) [inline src/nvim/fileio.c]: @mhinz what is the name `auwin_appended` trying to convey? That `win_alloc_aucmd_win` implicitly does `win_append`, so it should not be done again below?

File: src/nvim/fileio.c [hunk 4]
```diff
@@ -6520,21 +6520,14 @@ bool check_nomodeline(char_u **argp)
...
-/*
- * Prepare for executing autocommands for (hidden) buffer "buf".
- * Search for a visible window containing the current buffer.  If there isn't
- * one then use "aucmd_win".
- * Set "curbuf" and "curwin" to match "buf".
- */
-void 
-aucmd_prepbuf (
-    aco_save_T *aco,               /* structure to save values in */
-    buf_T *buf               /* new curbuf */
```
Comment (justinmk) [inline src/nvim/fileio.c]: I restored the comments as `@param` items in the merge-commit.

File: src/nvim/fileio.c (hunk 3 above)
Comment (mhinz) [reply src/nvim/fileio.c]: A belated "yes". :)

**Pull request merged**